File: Leetcode/0020 Valid Parentheses.ts
const rightToLeftMap = {
    ')': '(', 
    '}': '{',
    ']': '[',
};

function isValid(s: string): boolean {
    const stack = [];
    for(const char of s) {
        const isLeftParentheses = ['(', '{', '['].includes(char);
        if(isLeftParentheses) {
            stack.push(char);
            continue;
        }
        const stackTop = stack[stack.length - 1];
        if(stackTop == null || stackTop !== rightToLeftMap[char]) {
            return false;
        }
        stack.pop();
    }
    return stack.length === 0;
}

File: Leetcode/0020 Valid Parentheses.py
right_to_left_map = {
    ')': '(',
    '}': '{',
    ']': '['
}

class Solution:
    def isValid(self, s):
        stack = []
        for char in s:
            is_left_parentheses = char in ['(', '{', '[']
            if is_left_parentheses:
                stack.append(char)
                continue 
            if len(stack) == 0 or stack[-1] != right_to_left_map[char]:
                return False
            stack.pop()
        return len(stack) == 0
